blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,292 @@
+
+```racket
+(All (a) (-> (Vectorof a) Integer a))
+```
+
+It states that the index (the second argument) must be an `Integer`, i.e. any whole-number value in the exclusive range (-∞ , ∞). This description, however, is somewhat coarse: we obviously cannot use _any_ integer. To avoid a runtime error when indexing into a vector `v`, we must use an integer in the range [0, `(vector-length v)`). With a refinement type we can precisely describe this subset of the integers:
+
+```racket
+(Refine [i : Integer] (<= 0 i (- (vector-length v) 1)))
+```
+
+This refinement type describes values `i` of type `Integer` such that `(<= 0 i (- (vector-length v) 1))` produces `#true`. The documentation for [Refine](http://docs.racket-lang.org/search/index.html?q=Refine) describes precisely what propositions can be used in refinements. Today it includes logical statements about the types of terms, standard logical combinators (e.g. and, or), and linear integer constraints.
+
+## Dependent function types
+
+By combining refinement types with dependent function types (i.e. functions whose domain and range may depend on argument and/or program values), we can now give a more precise function type for many basic vector operations, e.g.:
+
+```racket
+(: safe-vector-ref 
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : (v) (Refine [i : Integer]
+                                  (<= 0 i (- (vector-length v) 1)))])
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Here we have stated that `safe-vector-ref`'s second argument (`n`) depends on its first argument (`v`) and that the type of `n` is an `Integer` which is a valid index for `v`.
+
+If we try and use `safe-vector-ref` with a vector and index which Typed Racket cannot prove satisfy the respective types:
+
+```racket
+(safe-vector-ref (vector 0 1) 2)
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+When a refinement on an argument's type may also be written conveniently as a "precondition", e.g.:
+
+```racket
+(: safe-vector-ref
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Natural])
+                #:pre (v n) (< n (vector-length v))
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Typed Racket may be able to provide more detailed feedback when the arguments are of the correct type but the precondition cannot be proved:
+
+```
+Type Checker: could not apply function;
+ unable to prove precondition
+  precondition: (<= 2 (- (vector-length v) 1)) 
+  in: (safe-vector-ref (vector 0 1) 2)
+```
+
+We can read the precondition specification above as follows: the precondition for the function depends are arguments `v` and `n`, and it requires that Typed Racket can prove `(<= 0 n (- (vector-length v) 1))` would produce `#true` at all call sites.
```

Comment:
Something is wrong after the colon.  Should it be "the precondition for the function depends on"?